File: php/utils/resolve_khinsider_url.php
<?php

function resolve_khinsider_url(string $url): ?string {
    $html = file_get_contents($url);
    if (!$html) {
        return null;
    }

    if (preg_match('/<audio.*?src="(.*?)".*?>/', $html, $matches)) {
        return $matches[1];
    }

    return null;
}

File: ruby/utils/resolve_khinsider_url.rb
# frozen_string_literal: true

require 'httparty'
require 'nokogiri'

module ResolveKhinsiderUrl
  def self.resolve(url)
    return url unless url.start_with?('https://downloads.khinsider.com/game-soundtracks')

    response = HTTParty.get(url)
    return url unless response.code == 200

    doc = Nokogiri::HTML(response.body)
    audio_player = doc.at_css('audio')
    audio_player ? audio_player['src'] : url
  end
end
